File: rabbitmq/cpp/receiver_async_test.cpp
#include <rabbitmq-c/amqp.h>
#include <rabbitmq-c/tcp_socket.h>
#include <iostream>
#include <string>
#include <signal.h>
#include <atomic>
#include <thread>
#include "../../utils/cpp/message_helpers.hpp"

using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

std::atomic<bool> running(true);

void signal_handler(int sig) {
    running = false;
}

int main(int argc, char* argv[]) {
    int receiver_id = 0;
    
    for (int i = 1; i < argc; i++) {
        if (std::string(argv[i]) == "--id" && i + 1 < argc) {
            receiver_id = std::stoi(argv[i + 1]);
            break;
        }
    }
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    amqp_connection_state_t conn = amqp_new_connection();
    amqp_socket_t *socket = amqp_tcp_socket_new(conn);
    amqp_socket_open(socket, "localhost", 5672);
    amqp_login(conn, "/", 0, 131072, 0, AMQP_SASL_METHOD_PLAIN, "guest", "guest");
    amqp_channel_open(conn, 1);

    std::string queue_name = "test_queue_" + std::to_string(receiver_id);
    amqp_queue_declare(conn, 1, amqp_cstring_bytes(queue_name.c_str()), 0, 0, 0, 0, amqp_empty_table);
    amqp_basic_consume(conn, 1, amqp_cstring_bytes(queue_name.c_str()), amqp_empty_bytes, 0, 1, 0, amqp_empty_table);

    std::cout << " [*] [ASYNC] Receiver " << receiver_id << " waiting for messages on " << queue_name << std::endl;

    while (running) {
        struct timeval timeout = {1, 0};
        amqp_envelope_t envelope;
        amqp_rpc_reply_t res = amqp_consume_message(conn, &envelope, &timeout, 0);

        if (res.reply_type == AMQP_RESPONSE_NORMAL) {
            std::string message_str((char*)envelope.message.body.bytes, envelope.message.body.len);
            
            MessageEnvelope msg_envelope;
            if (message_helpers::parse_envelope(message_str, msg_envelope)) {
                std::string message_id = msg_envelope.message_id();
                std::cout << " [x] [ASYNC] Received message " << message_id << std::endl;

                // Create ACK
                MessageEnvelope response = message_helpers::create_ack_from_envelope(
                    msg_envelope,
                    std::to_string(receiver_id)
                );
                response.set_async(true);
                std::string response_str = message_helpers::serialize_envelope(response);

                // Send ACK with proper binary handling
                amqp_basic_properties_t props;
                props._flags = AMQP_BASIC_CONTENT_TYPE_FLAG | AMQP_BASIC_CORRELATION_ID_FLAG;
                props.content_type = amqp_cstring_bytes("application/octet-stream");
                props.correlation_id = envelope.message.properties.correlation_id;

                // Use amqp_bytes_t to properly handle binary data (not null-terminated)
                amqp_bytes_t body;
                body.len = response_str.size();
                body.bytes = (void*)response_str.data();

                amqp_basic_publish(conn, 1, amqp_empty_bytes, envelope.message.properties.reply_to,
                                  0, 0, &props, body);
            }
            amqp_destroy_envelope(&envelope);
        }
    }

    std::cout << " [x] [ASYNC] Receiver " << receiver_id << " shutting down" << std::endl;
    amqp_channel_close(conn, 1, AMQP_REPLY_SUCCESS);
    amqp_connection_close(conn, AMQP_REPLY_SUCCESS);
    amqp_destroy_connection(conn);
    return 0;
}

File: rabbitmq/python/receiver_async_test.py
#!/usr/bin/env python3
"""RabbitMQ Python Receiver - Async"""
import sys
import signal
import asyncio
import aio_pika
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *

running = True

def signal_handler(sig, frame):
    global running
    running = False

async def run(receiver_id):
    connection = await aio_pika.connect_robust("amqp://guest:guest@localhost/")
    
    async with connection:
        channel = await connection.channel()
        queue_name = f"test_queue_{receiver_id}"
        queue = await channel.declare_queue(queue_name)
        
        print(f" [*] [ASYNC] Receiver {receiver_id} waiting for messages on {queue_name}")
        
        async with queue.iterator() as queue_iter:
            async for message in queue_iter:
                async with message.process():
                    request_envelope = parse_envelope(message.body)
                    message_id = request_envelope.message_id
                    print(f" [x] [ASYNC] Received message {message_id}")
                    
                    # Create ACK
                    response = create_ack_from_envelope(request_envelope, str(receiver_id))
                    setattr(response, 'async', True)
                    resp_str = serialize_envelope(response)
                    
                    # Send reply
                    await channel.default_exchange.publish(
                        aio_pika.Message(
                            body=resp_str,
                            correlation_id=message.correlation_id,
                            content_type='application/octet-stream'
                        ),
                        routing_key=message.reply_to
                    )
                
                if not running:
                    break

    print(f" [x] [ASYNC] Receiver {receiver_id} shutting down")


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--id', type=int, default=0)
    args = parser.parse_args()
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    # Loop setup
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    
    try:
        loop.run_until_complete(run(args.id))
    except KeyboardInterrupt:
        pass
    finally:
        loop.close()


if __name__ == "__main__":
    main()
